### PR TITLE
Remove 'using namespace std' from header files, properly namespace all std symbols

### LINKED ISSUE #1613: Towards enabling ninja to be built as a library
I wish to use ninja's code as a library inside of another application.

I'm well aware that ninja does not attempt to be a library, and am not asking for the Ninja project to support any kind of stable API.

I would like to remove the use of "using namespace std" in headerfiles, however, so that I don't have to deal with namespace pollution issues in my code that uses the ninja code as a library.

If I submit a pull request that does this, will it be accepted (assuming high enough quality) ?

### PR DESCRIPTION
**Pull request by jonesmz**
I wish to use ninja as a library in another application.

I am aware that the ninja codebase does not attempt to offer any kind of stable API, and I'm not looking for Ninja to do so in the future.

This pull request, if merged, would resolve #1613

My motivations for using the ninja code in a library situation include, but are not limited to, wanting to be able to parse the existing dependency database, and also the ninja log file, to drive some application specific behavior that ninja itself wouldn't benefit from.

There are a variety of other situations where supporting ninja in a library format would also make sense.

E.g. consider a situation where a program written in an interpreted language wanted to use ninja to execute some tasks. That application might want to modify the build graph while executing parts of it based on external information.

Right now, Ninja can't be used in that way, but I strongly believe that from both an architectural point of view, as well as a performance standpoint, moving the Ninja codebase toward a model where it's easier to use as a library (even if that's never officially supported!!!) will do nothing but benefit Ninja. Not the least reason of which is the ease of future maintenance of the code.

There are plenty more reasons, but I don't need to belabor the point.

Regardless, the problem that I continually have, is that it's not practical to directly #include headers from ninja into my existing project. I already *have* data types named map, and string, that live inside their own namespaces. But because ninja has "using namespace std", if I try to use "map" or "string" inside my namespace, but without the namespace indicator, I get ambiguity on which "map" or "string" is actually being used.

I know that Ninja uses "using namespace std" in it's code intentionally. I, of course, disagree with this, as have many others on the mailing list and on github.

So this pull request is a compromise.

Each implementation file can continue using, or not using, "using namespace std" as it sees fit, while header files will use "std::" when referring to types from the std namespace. This makes it significantly less difficult for downstream projects (Not only my own project of course, any downstream project at all) to use Ninja's codebase for their projects.

### CODE REVIEW

**Comment by jhasse** [comment]
Problem: This will result in lots of merge conflict in all open PRs (which there are a lot of unfortunately).

My idea would be to use `std::` in new code only or when touching a line anyway. Of course it will take a lot longer, before you can use the headers then :/

**Comment by jonesmz** [comment]
I don't want to sound unreasonable, but I honestly don't really see why merge conflicts with existing pull requests are a problem.

It should be very easy to update any existing pull requests  to work with this change, especially since it's something that can be done with a simple sed script.

Are there specific pull requests that you would like to point out? I am willing to update them, though I don't see why we should need to do so when many of the existing PRs already have (or used to last I checked) merge conflicts anyway.

If not, how can we plan to execute this change in a way you're willing to accept?

Can we identify specific header files that can be done right now, and then change other header files over time?

If you can give me a specific list, I can modify this pull request to only modify the indicated header files.

Note: The majority of files in this PR are only changed to add "using namespace std;" to the top of the cc files.

Some other proposal?

**Comment by jonesmz** [comment]
I looked through all of the open PRs, and approximately 50% of them have merge conflicts as is.

The reason for the merge conflicts appears to be primarily due to the age of these PRs.

I propose the Ninja project adopt the following policy:

PRs will be automatically closed if there is no activity (from anyone) for 3 months.

We can go through the current list of PRs and close any that haven't had any activity for that long.

Or alternatively, we can post a comment on each PR that has no activity for that time, to indicate the new policy is being enacted, and provide a one-time only grace period of 3 months before automatically being closed.

I also suggest the same policy be applied to the list of open issues, with nearly 200 open issues, it's a lot of noise to sift through.

**Comment by jhasse** [comment]
> Are there specific pull requests that you would like to point out?

#1600 for example.

> I am willing to update them, though I don't see why we should need to do so when many of the existing PRs already have (or used to last I checked) merge conflicts anyway.

Especially if they already have merge conflicts, we shouldn't make the situation worse!

> Can we identify specific header files that can be done right now, and then change other header files over time?

I don't know. If you want to help, feel free to review some of the open PRs (or if they have merge conflicts, just ask the creator to rebase).

> PRs will be automatically closed if there is no activity (from anyone) for 3 months.

Not a good idea IMHO. I have way too less time to look at most PRs, which doesn't mean it isn't planned to merge them at some point.

For issues, I would also like to keep them open as long as they aren't fixed (or WONTFIX).

**Comment by jhasse** [comment]
Replying here so that there isn't too much offtopic in #1600.

> Since this PR is **one of** the main justifications for delaying this PR #1614, it sounds like there is no longer **any** reason to wait to merge #1614.

See emphasis ;)
#1600 was just one example.

**Comment by jonesmz** [comment]
Help me to help you.

I asked for specific PRs you consider to be dependencies of this PR.

In the one PR you responded with, you said that it won't be merged any time soon, making any delay caused by the (relatively trivial) merge conflicts that would result into a non-issue.

If you have a specific list, give it to me, and I will communicate with the authors of those PRs if they have an objection to rebasing their PR after this is merged.

**Comment by jhasse** [comment]
I don't have a specific list and I don't have the time to create one, sorry. In general I'm trying to fix/merge these next: https://github.com/ninja-build/ninja/milestones/1.10.0

**Comment by jonesmz** [comment]
Continued:

Btw, I do apologize for my poor wording.

Please understand that I want to get things moving.

This PR represents a major hassle that I have to deal with every time Ninja merges code. I have to rebase this change every single time.

I want to stop dealing with that.

Coincidentally, this change also improves efforts toward #1499.

Can I assume that this PR will be merged when the list here is merged? https://github.com/ninja-build/ninja/milestones/1.10.0

File: src/test.h
Comment (jhasse) [inline src/test.h]: Can you clang-format this line (and the others with the same issue) so that it stays in the 80 columns limit?

**Comment by jhasse** [comment]
> Can I assume that this PR will be merged when the list here is merged? https://github.com/ninja-build/ninja/milestones/1.10.0

Yes, that sounds reasonable.

File: src/test.h
Comment (jonesmz) [reply src/test.h]: Yes. I will do so.

Comment (jonesmz) [reply src/test.h]: Clang formatting the code (using the supplied .clang-format file) results in *lots* of changes.

I suppose I'll edit things by hand then.

Comment (jonesmz) [reply src/test.h]: Actually.

Can we leave formatting changes, such as the 80 column limit, to a separate PR?

Comment (jhasse) [reply src/test.h]: Use Visual Studio Code (with C/C++ or ClangFormat extension). You can open the diff of each file and select the changed lines (on the right). Right-click on the selection and select "Format selection". It will only format the selected line then.

You can also add rulers so that you know where you're over the 80 limit to only format those.

Comment (jonesmz) [reply src/test.h]: I suspect you were writing your reply while I asked:

Can we leave formatting changes, such as the 80 column limit, to a separate PR?

Comment (jonesmz) [reply src/test.h]: I suggest that .clang-format be run on the whole codebase after https://github.com/ninja-build/ninja/milestones/1.10.0

Comment (jhasse) [reply src/test.h]: No, sorry. I would like to avoid line touches just for formatting.

Comment (jonesmz) [reply src/test.h]: You understand that the diff between what we have, and what clang-format gives us is over 5,000 lines right?

clang-format -i src/*.h src/*.cc ; git diff | grep -e "+" -e "-" |  wc -l
8808

You need to rip the bandaid off eventually.

It would be better for *everyone* if we just applied clang format, and moved on.

Comment (jhasse) [reply src/test.h]: I disagree.

Comment (jonesmz) [reply src/test.h]: The second commit on this PR addresses lines changed by the first commit that now exceed 80 columns.

Comment (jonesmz) [reply src/test.h]: > I disagree.

Can you please provide explanations in the future? No meaningful discussion can be had if one party provides explanations for their position, e.g. The clang formatting rules supplied with the project cause over 5 thousand lines of changes

when the other person just says they disagree and that's all.

Comment (jhasse) [reply src/test.h]: You didn't explain, why it would be better for everyone, if we just applied clang-format and moved on, either. That's why I felt the need to state, that I disagree with that opinion.

Comment (jonesmz) [reply src/test.h]: Ok, I can see how you thought that.

I thought it was implied by "There are over 5 thousand lines of changes", and the work that contributors have to do by hand in order to comply with the not-in-clang-format rules that the project has.

E.g., it took me 20+ minutes to go line by line and correct the overly long lines in a diff viewer.

Why bother to offer a clang format ruleset if it's not used?

If we apply the rules to the entire repository, and then consistently enforce the use of clang format for any PR that's merged (either by requiring the PR be formatted with clang format prior to merging, or applying clang format immediately after merging), the pain of requiring each contributor to comply individually is condensed into a single "flag day", and then goes away after that.

Comment (jhasse) [reply src/test.h]: Would have took you only a few minutes if you done it with Visual Studio Code.

> Why bother to offer a clang format ruleset if it's not used?

There's not clang-format ruleset. If in doubt -> use clang-format.

Comment (jonesmz) [reply src/test.h]: > Would have took you only a few minutes if you done it with Visual Studio Code.

This would require:
1) I have Visual Studio Code, or am able to install it. You don't know what operating system I use, or if it's practical for me to download and install an arbitrary program.
2) That Visual Studio Code is, in fact, faster to use than whatever tool I'm already familiar with.

> There's not clang-format ruleset. If in doubt -> use clang-format.

.clang-format == "clang-format ruleset"

Comment (jhasse) [reply src/test.h]: > That Visual Studio Code is, in fact, faster to use than whatever tool I'm already familiar with.

Apparently it is, otherwise it wouldn't have took you 20 minutes.

Comment (jonesmz) [reply src/test.h]: I did a lot more than just re-flow 80+column lines. I compared the clang formatting of every line in the header files to make sure it all looked right.

Visual studio code wouldn't have helped much with that.

Comment (jhasse) [reply src/test.h]: Ah I see. No need to do that :)

**Comment by jonesmz** [comment]
Anything else this PR needs before merging?

**Comment by jhasse** [comment]
I (or someone else) needs to review it.

**Comment by jonesmz** [comment]
Anything I'm able to do to accelerate this?

**Comment by jhasse** [comment]
Fix #1724 for example.

**Comment by jonesmz** [comment]
What does that have to do with a mechanical, code only change?

**Comment by jhasse** [comment]
I will fix #1724 prior to this PR. If someone else fixes #1724 before me, I'll have time to look at this PR.

**Comment by jonesmz** [comment]
Thank you for clarifying.

**Comment by jonesmz** [comment]
Its now been an additional 3 months.

This is a mechanical only change that has no realistic chance of causing problems.

Can we please just merge it? Every time another PR is merged, I have to re-rebase this commit and all of my other patches that depend on it (both patches submitted here as PRs as well as patches elsewhere)

**Comment by jonesmz** [comment]
Another month, and now I've had to rebase the changes to fix merge conflicts twice.

Can we please just merge this PR? It changes no functionality. There's nothing to review!

**Comment by jhasse** [comment]
I'm sorry and feel your frustration. But #1724 still hasn't been fixed and I could definitely use some help there.

**Comment by jonesmz** [comment]
> But #1724 still hasn't been fixed and I could definitely use some help there.

I'm not sympathetic to that bug, so I have no interest in contributing to fixing it.

This PR trivial to review, makes no changes in runtime functionality, and not having it merged is making my life more difficult.

Please just merge it.

**Comment by jonesmz** [comment]
Rebased again to account for header files that were modified since this PR was created that were being missed until now.

Additionally: I split out the main commit (not the 80-column commit) into 2.

1. Add 'using namespace std;' to .cc files.
2. Remove 'using namespace std;' from .h files, and fix the resulting compiler errors.

This should make the PR easier to review.

**Comment by jonesmz** [comment]
Another month

At the very least merge the first commit. Its fucking tiny.

This entire PR is tiny.

**Comment by jonesmz** [comment]
It's been another month, and now #1724 is merged / resolved. Aside from a single comment being added, there were no merge conflicts with that change, so why did this PR need to wait for that one?

Can you please merge this now?

**Comment by jhasse** [comment]
I have explained that in another PR: Because that was the issue that was on the top of my todo list and I rarely have enough time to put in for other stuff.

Currently I need to do the 1.10.1 release next. Please stop bugging me.

**Comment by jonesmz** [comment]
This is my last message about this PR, and potentially my last interaction with Ninja as a project entirely.

I can't believe that you've seriously let such a trivial PR sit for over a year. It's degrading to contributors and seriously diminishes the credibility of the Ninja project as code that can be relied on.

Your inability to handle the basic maintenance tasks for this project in a timely manner have serious ramifications for contributors across the world. I personally experienced not one, but two, projects I was involved in at work get cancelled *specifically* because of this project's maintainers inability to handle trivial changes.

If the maintainers of Ninja are unable to do simple things in a timely manner, find someone who can.

**Comment by jhasse** [comment]
It's hard to find good people for that salary.

File: src/build.h
Comment (jhasse) [inline src/build.h:261]: fix alignment

File: src/clparser.h
Comment (jhasse) [inline src/clparser.h:30]: alignment

File: src/eval_env.h
Comment (jhasse) [inline src/eval_env.h:32]: I think this was meant more as a general string, not a std::string in particular.

File: src/eval_env.h
Comment (jhasse) [inline src/eval_env.h:101]: alignment

File: src/graph.h
Comment (jhasse) [inline src/graph.h:82]: alignment

File: src/includes_normalize.h
Comment (jhasse) [inline src/includes_normalize.h:28]: Did clang-format do this? If not, please don't.

File: src/manifest_parser.h
Comment (jhasse) [inline src/manifest_parser.h:53]: remove std:: here

File: src/string_piece.h
Comment (jhasse) [inline src/string_piece.h:22]: also don't mention std:: here

File: src/string_piece.h
Comment (jhasse) [inline src/string_piece.h:44]: You've added two std:: here.

File: src/string_piece.h
Comment (jhasse) [inline src/string_piece.h:45]: keep string without std::

**Review by jhasse — CHANGES_REQUESTED** [review]
Have you checked if we can enable https://clang.llvm.org/extra/clang-tidy/checks/google-global-names-in-headers.html now?

**Pull request merged**
